File: 3. Behavioral Design Patterns/Strategy/C#/StrategyExample.cs
// Strategy Pattern Example in C#
// Sorter context with BubbleSort and QuickSort strategies

using System;
using System.Collections.Generic;

namespace StrategyExample
{
    interface ISortStrategy { IList<int> Sort(IList<int> data); }

    class BubbleSort : ISortStrategy
    {
        public IList<int> Sort(IList<int> data)
        {
            var arr = new List<int>(data);
            for (int i = 0; i < arr.Count; i++)
                for (int j = 0; j < arr.Count - i - 1; j++)
                    if (arr[j] > arr[j + 1]) { (arr[j], arr[j + 1]) = (arr[j + 1], arr[j]); }
            Console.WriteLine("BubbleSort used");
            return arr;
        }
    }

    class QuickSort : ISortStrategy
    {
        public IList<int> Sort(IList<int> data)
        {
            var arr = new List<int>(data);
            arr.Sort(); // use built-in for brevity
            Console.WriteLine("QuickSort used");
            return arr;
        }
    }

    class Sorter
    {
        private ISortStrategy _strategy;
        public Sorter(ISortStrategy strategy) { _strategy = strategy; }
        public void SetStrategy(ISortStrategy strategy) { _strategy = strategy; }
        public IList<int> Sort(IList<int> data) => _strategy.Sort(data);
    }

    class Program
    {
        static void Main()
        {
            var data = new List<int> { 5, 2, 9, 1, 5, 6 };
            var sorter = new Sorter(new BubbleSort());
            Console.WriteLine(string.Join(", ", sorter.Sort(data)));
            sorter.SetStrategy(new QuickSort());
            Console.WriteLine(string.Join(", ", sorter.Sort(data)));
        }
    }
}

File: 3. Behavioral Design Patterns/Strategy/Python/StrategyExample.py
# Strategy Pattern Example in Python
# Sorter context with BubbleSort and QuickSort strategies

from abc import ABC, abstractmethod

class SortStrategy(ABC):
    @abstractmethod
    def sort(self, data):
        pass

class BubbleSort(SortStrategy):
    def sort(self, data):
        arr = data.copy()
        n = len(arr)
        for i in range(n):
            for j in range(0, n - i - 1):
                if arr[j] > arr[j + 1]:
                    arr[j], arr[j + 1] = arr[j + 1], arr[j]
        print("BubbleSort used")
        return arr

class QuickSort(SortStrategy):
    def sort(self, data):
        arr = data.copy()
        arr.sort()  # use built-in for brevity
        print("QuickSort used")
        return arr

class Sorter:
    def __init__(self, strategy):
        self.strategy = strategy
    
    def set_strategy(self, strategy):
        self.strategy = strategy
    
    def sort(self, data):
        return self.strategy.sort(data)

def main():
    data = [5, 2, 9, 1, 5, 6]
    sorter = Sorter(BubbleSort())
    print(sorter.sort(data))
    sorter.set_strategy(QuickSort())
    print(sorter.sort(data))

if __name__ == "__main__":
    main()
